File: sdk-extras/src/test/java/eu/darken/porter/sdk/extras/FakeExtrasService.java
package eu.darken.porter.sdk.extras;

import android.content.pm.PackageManager;
import android.os.Bundle;
import android.os.IBinder;

import eu.darken.porter.server.IPorterApplication;
import eu.darken.porter.server.IPorterRemoteProcess;
import eu.darken.porter.server.IPorterService;
import eu.darken.porter.server.IPorterServiceConnection;

/** A local stub standing in for the server, carrying only the system property calls. */
class FakeExtrasService extends IPorterService.Stub {

    Bundle attachReply = new Bundle();

    /** Answered to every {@code getSystemProperty}; null means "echo the default back". */
    String propertyValue;

    String readName;
    String readDefault;
    String writtenName;
    String writtenValue;

    @Override
    public Bundle attach(IPorterApplication application, Bundle args) {
        return attachReply;
    }

    @Override
    public String getSystemProperty(String name, String defaultValue) {
        readName = name;
        readDefault = defaultValue;
        return propertyValue != null ? propertyValue : defaultValue;
    }

    @Override
    public void setSystemProperty(String name, String value) {
        writtenName = name;
        writtenValue = value;
    }

    @Override
    public int getUid() {
        return -1;
    }

    @Override
    public int checkPermission(String permission) {
        return PackageManager.PERMISSION_DENIED;
    }

    @Override
    public String getSELinuxContext() {
        return null;
    }

    @Override
    public IPorterRemoteProcess newProcess(String[] cmd, String[] env, String dir) {
        return null;
    }

    @Override
    public int addUserService(IPorterServiceConnection conn, Bundle args) {
        return 0;
    }

    @Override
    public int removeUserService(IPorterServiceConnection conn, Bundle args) {
        return 0;
    }

    @Override
    public void requestPermission(int requestCode) {
    }

    @Override
    public boolean checkSelfPermission() {
        return false;
    }

    @Override
    public boolean shouldShowRequestPermissionRationale() {
        return false;
    }

    @Override
    public void exit() {
    }

    @Override
    public void attachUserService(IBinder binder, Bundle args) {
    }

    @Override
    public void dispatchPermissionConfirmationResult(
            int requestUid, int requestPid, int requestCode, Bundle data) {
    }

    @Override
    public int getFlagsForUid(int uid, int mask) {
        return 0;
    }

    @Override
    public void updateFlagsForUid(int uid, int mask, int value) {
    }
}

File: sdk-extras/src/test/java/eu/darken/porter/sdk/extras/FakeExtrasService.kt
package eu.darken.porter.sdk.extras

import android.content.pm.PackageManager
import android.os.Bundle
import android.os.IBinder
import eu.darken.porter.server.IPorterApplication
import eu.darken.porter.server.IPorterRemoteProcess
import eu.darken.porter.server.IPorterService
import eu.darken.porter.server.IPorterServiceConnection

/** A local stub standing in for the server, carrying only the system property calls. */
internal class FakeExtrasService : IPorterService.Stub() {

    var attachReply: Bundle = Bundle()

    /** Answered to every `getSystemProperty`; null means "echo the default back". */
    var propertyValue: String? = null

    var readName: String? = null
    var readDefault: String? = null
    var writtenName: String? = null
    var writtenValue: String? = null

    override fun attach(application: IPorterApplication, args: Bundle): Bundle = attachReply

    override fun getSystemProperty(name: String, defaultValue: String?): String? {
        readName = name
        readDefault = defaultValue
        return propertyValue ?: defaultValue
    }

    override fun setSystemProperty(name: String, value: String) {
        writtenName = name
        writtenValue = value
    }

    override fun getUid(): Int = -1

    override fun checkPermission(permission: String): Int = PackageManager.PERMISSION_DENIED

    override fun getSELinuxContext(): String? = null

    override fun newProcess(cmd: Array<String>?, env: Array<String>?, dir: String?): IPorterRemoteProcess? = null

    override fun addUserService(conn: IPorterServiceConnection?, args: Bundle): Int = 0

    override fun removeUserService(conn: IPorterServiceConnection?, args: Bundle): Int = 0

    override fun requestPermission(requestCode: Int) {
    }

    override fun checkSelfPermission(): Boolean = false

    override fun shouldShowRequestPermissionRationale(): Boolean = false

    override fun exit() {
    }

    override fun attachUserService(binder: IBinder, args: Bundle) {
    }

    override fun dispatchPermissionConfirmationResult(requestUid: Int, requestPid: Int, requestCode: Int, data: Bundle) {
    }

    override fun getFlagsForUid(uid: Int, mask: Int): Int = 0

    override fun updateFlagsForUid(uid: Int, mask: Int, value: Int) {
    }
}
